doc/tutorials/run_all_tests_with_docker.md

```diff
@@ -0,0 +1,132 @@
+
+If you want to view all the available images, execute this command:
+
+```sh
+docker run --rm anoxis/registry-cli -r https://hub-public.libelektra.org
+```
+
+You will see something like this:
+
+```
+---------------------------------
+Image: build-elektra-debian-stretch
+  tag: 201906-ecf9161f41a8b472b3b0282a85a9f91d1f0f45357756e5451ae043fce8d0100e
+  tag: 201902-b6d49f470e1171348248b2f87ef397d58d7a2dae14d201f4073564079ce0c070
+  tag: 201903-b95dc56352aa684e16dfb8628bded4c69c712223f5d7ed99ebdd644852a32123
+  tag: 201905-ecf9161f41a8b472b3b0282a85a9f91d1f0f45357756e5451ae043fce8d0100e
+  tag: 201901-6b08855f13ba26e3ad1fa80e399b87df860cc24889f2d1854fa0050834567b26
+  tag: 201904-ecf9161f41a8b472b3b0282a85a9f91d1f0f45357756e5451ae043fce8d0100e
+  tag: 201904-1a6be7b9c3740a2338b14d08c757332cae5254ce58219b6cc2908c7bd6e4f460
+  tag: 201903-1a6be7b9c3740a2338b14d08c757332cae5254ce58219b6cc2908c7bd6e4f460
+  tag: 201903-6b08855f13ba26e3ad1fa80e399b87df860cc24889f2d1854fa0050834567b26
+  tag: 201902-6b08855f13ba26e3ad1fa80e399b87df860cc24889f2d1854fa0050834567b26
+```
+
+Afterwards pull your desired image as you would do from any public registry:
+
+```sh
+docker pull hub-public.libelektra.org/<image_name>:<tag_name>
+```
+
+Example:
+
+```sh
+docker pull hub-public.libelektra.org/build-elektra-debian-stretch:201905-9dfe329fec01a6e40972ec4cc71874210f69933ab5f9e750a1c586fa011768ab
+```
+
+### 2. Run the Docker Container
+
+You have to be in the root of the Elektra project, so that the container can properly map all the source files.
+
+So from your root project folder run the following:
+
+```sh
+docker run -it --rm \
+-v "$PWD:/home/jenkins/workspace" \
+-w /home/jenkins/workspace \
+hub-public.libelektra.org/<image_name>:<tag_name>
+```
+
+Example:
+
+```sh
+docker run -it --rm \
+-v "$PWD:/home/jenkins/workspace" \
+-w /home/jenkins/workspace \
+ hub-public.libelektra.org/build-elektra-debian-stretch:201905-9dfe329fec01a6e40972ec4cc71874210f69933ab5f9e750a1c586fa011768ab
+```
+
+### 3. Build
+
+After starting the container, you should be automatically inside it in the working directory `/home/jenkins/workspace`.
+
+Create folder for building the project and `cd` to it like this:
+
+```sh
+mkdir build-docker && cd build-docker
+```
+
+Build it with
+
+```sh
+ cmake .. \
+-DBINDINGS="ALL;-DEPRECATED;-haskell" \
+-DPLUGINS="ALL;-DEPRECATED" \
+-DTOOLS="ALL" \
+-DENABLE_DEBUG=ON \
+-DKDB_DB_HOME="$PWD" \
+-DKDB_DB_SYSTEM="$PWD/.config/kdb/system" \
+-DKDB_DB_SPEC="$PWD/.config/kdb/system"
+```
+
+and then with
+
+```sh
+make -j 10
+```
+
+Additionally, you may want to install Elektra inside the container, because the installed tests rely on it.
+The build server also does this and it is possible that the installed tests have different results (e.g. if test data is missing)
+Just run this command:
+
+```sh
+make install
```

Comment:
This step fail on my machine with the error:

```
-- Install configuration: "RelWithDebInfo"
CMake Error at cmake_install.cmake:36 (file):
  file cannot create directory: /usr/local/share/doc/elektra.  Maybe need
  administrative privileges.


Makefile:72: recipe for target 'install' failed
make: *** [install] Error 1
```

. It might make sense to add the option `-DINSTALL_SYSTEM_FILES=OFF` to the `cmake` command above.

Reply:
Thank you for your comment! Now I am getting this error too... Unfortunately, adding -DINSTALL_SYSTEM_FILES=OFF did not help. Does anyone have a suggestion on how to solve it? @markus2330

Reply:
I think setting `CMAKE_INSTALL_PREFIX` to a newly created folder inside the repository (e.g. `"$PWD/install"`) should help.

Reply:
I am sorry, but I don't understand what do you mean. Can you make an example?

Reply:
> Can you make an example?

Just replace the current `cmake` command with:


```sh
cmake .. \
 -DBINDINGS="ALL;-DEPRECATED;-haskell" \
 -DPLUGINS="ALL;-DEPRECATED" \
 -DTOOLS="ALL" \
 -DENABLE_DEBUG=ON \
 -DKDB_DB_HOME="$PWD" \
 -DKDB_DB_SYSTEM="$PWD/.config/kdb/system" \
 -DKDB_DB_SPEC="$PWD/.config/kdb/system" \
 -DCMAKE_INSTALL_PREFIX="$PWD/install" \
 -DINSTALL_SYSTEM_FILES=OFF
```

, and installing should work.

Reply:
Thank you! It works indeed. Additionally, I had to add the installation to the PATH like this

`export PATH="$PWD/install/bin:$PATH"`

Even though the installation works now and I can interact with the kdb, these tests are still failing for some reason. Do you have an idea why? 

 ```
99 - testshell_markdown_file (Failed)
        230 - testshell_markdown_tutorial_cascading (Failed)
        232 - testshell_markdown_tutorial_validation (Failed)
        276 - testkdb_ensure (Failed)

```
Detailed output:

```
kdb set /tests/tutorial/spec/should_not_be_here abc

ERROR - RET:
Return value “0” does not match “5”


ERROR - ERROR:
“”
does not match
“10”

kdb get /tests/tutorial/spec/should_not_be_here

ERROR - RET:
Return value “0” does not match “11”


ERROR - STDERR:
“ Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial/spec/should_not_be_here not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint”
does not match
“Did not find key '/tests/tutorial/spec/should_not_be_here'”

kdb rm -r spec/tests/tutorial
kdb rm -r system/tests/tutorial

ERROR - RET:
Return value “1” does not match “0”

kdb rm -rf user/tests/tutorial
kdb umount spec/tests/tutorial
kdb umount /tests/tutorial
kdb rm -rf spec
kdb rm -rf user
kdb import spec dump < $(kdb get system/tests/specbackup)

ERROR - RET:
Return value “2” does not match “0”

kdb import user dump < $(kdb get system/tests/userbackup)

ERROR - RET:
Return value “2” does not match “0”

rm $(kdb get system/tests/specbackup)

ERROR - RET:
Return value “1” does not match “0”

rm $(kdb get system/tests/userbackup)

ERROR - RET:
Return value “1” does not match “0”

kdb rm system/tests/specbackup

ERROR - RET:
Return value “1” does not match “0”

kdb rm system/tests/userbackup

ERROR - RET:
Return value “1” does not match “0”

shell_recorder /tmp/tmp.RY96VWsAQw RESULTS: 83 test(s) done 11 error(s).

—— Protocol ————————————————————————————————————————————————————
CMD: kdb mount validation.dump user/tests/together dump validation
RET: 0

CMD: kdb vset user/tests/together/test 123 "[1-9][0-9]*" "Not a number"
RET: 0

CMD: kdb set user/tests/together/test abc
RET: 5
STDERR: Sorry, module validator issued the error 42:
key value failed to validate: Not a number
ERROR: 42

CMD: kdb lsmeta user/tests/together/test
RET: 0
STDOUT: check/validation
check/validation/match
check/validation/message

CMD: kdb setmeta user/tests/together/test check/validation "[1-9][0-9]*"
RET: 0

CMD: kdb setmeta user/tests/together/test check/validation/match LINE
RET: 0

CMD: kdb setmeta user/tests/together/test check/validation/message "Not a number"
RET: 0

CMD: kdb set user/tests/together/test 123
RET: 0
STDOUT: Set string to "123"

CMD: kdb rm -r user/tests/together
RET: 0

CMD: kdb umount user/tests/together
RET: 0

CMD: kdb set system/tests/specbackup $(mktemp)
RET: 0
STDOUT: Create a new key system/tests/specbackup with string "/tmp/tmp.jXSBb2jNjC"

CMD: kdb set system/tests/userbackup $(mktemp)
RET: 0
STDOUT: Create a new key system/tests/userbackup with string "/tmp/tmp.D7rD3fHBuC"

CMD: kdb export spec dump > $(kdb get system/tests/specbackup)
RET: 0
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/specbackup not belonging to "" with name "default" because it is not system
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/userbackup not belonging to "" with name "default" because it is not system
 Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/specbackup not belonging to "" with name "default" because it is not system
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/userbackup not belonging to "" with name "default" because it is not system

CMD: kdb export user dump > $(kdb get system/tests/userbackup)
RET: 0

CMD: kdb setmeta spec/tests/spec/test hello world
RET: 0
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/specbackup not belonging to "" with name "default" because it is not system
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/userbackup not belonging to "" with name "default" because it is not system

CMD: kdb set /tests/spec/test value
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: Using name user/tests/spec/test
Create a new key user/tests/spec/test with string "value"

CMD: kdb lsmeta spec/tests/spec/test | grep -v '^internal/ini'
RET: 0
STDOUT: hello

CMD: kdb lsmeta /tests/spec/test | grep -v '^internal/ini'
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: hello

CMD: kdb getmeta /tests/spec/test hello
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: world

CMD: kdb getmeta user/tests/spec/test hello || kdb getmeta system/tests/spec/test hello
RET: 0
STDOUT: world

CMD: kdb setmeta "spec/tests/spec/_" new metaval
RET: 0

CMD: kdb set /tests/spec/test value
RET: 0
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/_ not belonging to "" with name "default" because it is not spec
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: Using name user/tests/spec/test
Set string to "value"

CMD: kdb lsmeta /tests/spec/test | grep -v '^internal/ini'
RET: 0
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/_ not belonging to "" with name "default" because it is not spec
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: hello
new

CMD: kdb rm -r spec/tests/spec
RET: 0

CMD: kdb rm -r user/tests/spec || kdb rm -r system/tests/spec
RET: 0

CMD: kdb setmeta spec/tests/spec/test check/validation "[1-9][0-9]*"
RET: 0

CMD: kdb setmeta spec/tests/spec/test check/validation/match LINE
RET: 0

CMD: kdb setmeta spec/tests/spec/test check/validation/message "Not a number"
RET: 0

CMD: kdb set /tests/spec/test "not a number"
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: Using name user/tests/spec/test
Create a new key user/tests/spec/test with string "not a number"

CMD: kdb lsmeta /tests/spec/test | grep -v '^internal/ini'
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/spec/test not belonging to "" with name "default" because it is not spec
STDOUT: check/validation
check/validation/match
check/validation/message

CMD: kdb rm -r spec/tests/spec/test
RET: 0

CMD: kdb rm -r user/tests/spec || kdb rm -r system/tests/spec
RET: 0

CMD: kdb mount tutorial.dump spec/tests/tutorial dump
RET: 0

CMD: cat << HERE | kdb import spec/tests/tutorial ni  
[]                                         
 mountpoint=tutorial.dump                
 infos/plugins=dump validation           
[/links/_]                                 
check/validation=https?://.*\..*         
check/validation/match=LINE              
check/validation/message=not a valid URL 
description=A link to some website       
HERE
RET: 0

CMD: kdb lsmeta spec/tests/tutorial
RET: 0
STDOUT: infos/plugins
mountpoint

CMD: kdb spec-mount /tests/tutorial
RET: 0

CMD: kdb set /tests/tutorial/links/url "invalid url"
RET: 5
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial/links/_ not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
Sorry, module validator issued the error 42:
key value failed to validate: not a valid URL
STDOUT: Using name user/tests/tutorial/links/url
ERROR: 42

CMD: kdb file user/tests/tutorial
RET: 0
STDOUT: /home/jenkins/.config/tutorial.dump

CMD: kdb file system/tests/tutorial
RET: 0
STDOUT: /home/jenkins/workspace/build-docker/.config/kdb/system/tutorial.dump

CMD: kdb file dir/tests/tutorial
RET: 0
STDOUT: /home/jenkins/workspace/.dir/tutorial.dump

CMD: kdb set -N system /tests/tutorial/links/elektra https://www.libelektra.org
RET: 0
STDERR:  Sorry, 2 warnings were issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial/links/_ not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
STDOUT: Using name system/tests/tutorial/links/elektra
Create a new key system/tests/tutorial/links/elektra with string "https://www.libelektra.org"

CMD: kdb setmeta /tests/tutorial/spec/should_not_be_here trigger/error 10
RET: 0
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key system/tests/tutorial/links/elektra not belonging to "/tests/tutorial" with name "/tests/tutorial" but instead to "spec/tests/tutorial" with name "spec/tests/tutorial" because it is hidden by other mountpoint
STDOUT: Using keyname spec/tests/tutorial/spec/should_not_be_here

CMD: kdb spec-mount /tests/tutorial
RET: 0
STDOUT: No metadata "mountpoint" found below spec/tests/tutorial

CMD: kdb set /tests/tutorial/spec/should_not_be_here abc
RET: 0
=== FAILED return value does not match expected pattern 5
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial/spec/should_not_be_here not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
STDOUT: Using name user/tests/tutorial/spec/should_not_be_here
Create a new key user/tests/tutorial/spec/should_not_be_here with string "abc"
=== FAILED Errors do not match expected pattern 10

CMD: kdb get /tests/tutorial/spec/should_not_be_here
RET: 0
=== FAILED return value does not match expected pattern 11
STDERR:  Sorry, 1 warning was issued ;(
        Sorry, module  issued the warning 79:
        postcondition of backend was violated: drop key spec/tests/tutorial/spec/should_not_be_here not belonging to "spec/tests/tutorial" with name "spec/tests/tutorial" but instead to "/tests/tutorial" with name "/tests/tutorial" because it is hidden by other mountpoint
=== FAILED stderr does not match expected pattern Did not find key '/tests/tutorial/spec/should_not_be_here'
STDOUT: abc

CMD: kdb rm -r spec/tests/tutorial
RET: 0

CMD: kdb rm -r system/tests/tutorial
RET: 1
=== FAILED return value does not match expected pattern 0
STDERR: Did not find any key

CMD: kdb rm -rf user/tests/tutorial
RET: 0

CMD: kdb umount spec/tests/tutorial
RET: 0

CMD: kdb umount /tests/tutorial
RET: 0

CMD: kdb rm -rf spec
RET: 0
STDERR: Did not find any key

CMD: kdb rm -rf user
RET: 0
STDERR: Did not find any key

CMD: kdb import spec dump < $(kdb get system/tests/specbackup)
RET: 2
=== FAILED return value does not match expected pattern 0
STDERR: Did not find key 'system/tests/specbackup'sh: 1: cannot open : No such file

CMD: kdb import user dump < $(kdb get system/tests/userbackup)
RET: 2
=== FAILED return value does not match expected pattern 0
STDERR: Did not find key 'system/tests/userbackup'sh: 1: cannot open : No such file

CMD: rm $(kdb get system/tests/specbackup)
RET: 1
=== FAILED return value does not match expected pattern 0
STDERR: Did not find key 'system/tests/specbackup'rm: missing operand
Try 'rm --help' for more information.

CMD: rm $(kdb get system/tests/userbackup)
RET: 1
=== FAILED return value does not match expected pattern 0
STDERR: Did not find key 'system/tests/userbackup'rm: missing operand
Try 'rm --help' for more information.

CMD: kdb rm system/tests/specbackup
RET: 1
=== FAILED return value does not match expected pattern 0
STDERR: Did not find the key

CMD: kdb rm system/tests/userbackup
RET: 1
=== FAILED return value does not match expected pattern 0
STDERR: Did not find the key
————————————————————————————————————————————————————————————————

        Start 276: testkdb_ensure
  4/277 Test #276: testkdb_ensure ................................***Failed    0.22 sec
Running main() from /opt/gtest/googletest/src/gtest_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from Ensure
[ RUN      ] Ensure.GlobalUnmount
[       OK ] Ensure.GlobalUnmount (53 ms)
[ RUN      ] Ensure.Unmount
/home/jenkins/workspace/tests/kdb/testkdb_ensure.cpp:102: Failure
Expected equality of these values:
  root.getMeta<std::string> ("warnings/#00/number")
    Which is: "79"
  "3"
error plugin didn't run
[  FAILED  ] Ensure.Unmount (88 ms)
[ RUN      ] Ensure.GlobalMount
tracer: open(0x559236e860e0, spec/tests/kdb/ensure = ): 0
tracer: close(0x559236e860e0,  = )
[       OK ] Ensure.GlobalMount (69 ms)
[----------] 3 tests from Ensure (210 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (210 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Ensure.Unmount

 1 FAILED TEST
```

Reply:
> Do you have an idea why?

Nope. As I wrote before, most of those tests also failed on my machine (macOS). Can you please tell me which host OS you use?

Reply:
I am using Ubuntu 18.04 in a virtual machine.

Reply:
> I am using Ubuntu 18.04 in a virtual machine.

Interesting. Would it not make more sense to just use Docker directly? Anyway, it would be great, if someone who uses Linux as main OS, could check if the tests also fail on their machine.

Reply:
I could not install Elektra on Windows, so I decided to work entirely in a Linux virtual machine.

Reply:
> I could not install Elektra on Windows, so I decided to work entirely in a Linux virtual machine.

What did you try? Can you write an issue about that please?

> Interesting. Would it not make more sense to just use Docker directly?

Yes, I also wonder why you do not use docker directly?

>  Anyway, it would be great, if someone who uses Linux as main OS, could check if the tests also fail on their machine.

I can reproduce the same problem (3 failed tests:testshell_markdown_tutorial_cascading  testshell_markdown_tutorial_validation testkdb_ensure). Something is tidy inside the docker image.

Please report these three issues separately, they seems to be unrelated to this tutorial.

@sanssecours any idea why our build server does not catch this problem?

Reply:
> @sanssecours any idea why our build server does not catch this problem?

Nope.